Refactor/116 RegistryInfoContent is converted to functional component

agar23 opened the pull request:
<!--

Before Pull Request check whether your commits follow this convention

https://github.com/verdaccio/verdaccio/blob/master/CONTRIBUTING.md#git-commit-guidelines

  * If your PR fix an issue don't forget to update the unit test and documentation in /docs folder
  * If your PR delivers a new feature, please, provide examples and why such feature should be considered.
  * Document your changes /docs
  * Add unit test
  * Follow the commit guidelines in order to get a quick approval

Pick one/multiple type, if none apply please suggest one, we might be included it by default

eg: bug / feature / documentation / unit test / build

-->
**Type:** Refactor

The following has been addressed in the PR:

*  There is a related issue? #116 
*  Unit or Functional tests are included in the PR: Yes

**Description:**

Refactor RegistryInfoContent compoenent to functional

<!-- Resolves #??? -->

agar23 commented:
@priscilawebdev @juanpicado  please review.

priscilawebdev commented on src/components/Header/HeaderInfoDialog.tsx:
```diff
@@ -1,7 +1,7 @@
 
 import RegistryInfoDialog from '../RegistryInfoDialog';
-import RegistryInfoContent from '../RegistryInfoContent';
+import { RegistryInfoContent } from '../RegistryInfoContent';
```
Hmm... I would rather leave it as a default import as it it the main code of `RegistryInfoContent`

agar23 replied:
generally speaking, using default imports/exports can lead to issues. Is there a reason why we prefer to use them vs named imports/exports?

https://basarat.gitbooks.io/typescript/docs/tips/defaultIsBad.html
https://stackoverflow.com/a/33307487

priscilawebdev replied:
@agar23 thank you for share these info with us... I believe that it's something we can discuss for the upcoming verdaccio version.. I  will really read more about it, but please could you leave it as default for now? Then if we decide to do as you suggested,  we can add this change in all components 😉

agar23 replied:
sounds good 😄

priscilawebdev commented on src/components/PackageList/PackageList.tsx:
```diff
@@ -1,4 +1,4 @@
-import * as React from 'react';
+import React, { Fragment, ReactElement } from 'react';
```
Maybe we should discuss about some standards, but I am trying to keep the imports together, like you did here : 
![image](https://user-images.githubusercontent.com/29228205/67668022-6bfdec00-f96f-11e9-9eca-f184947f8bf5.png)

agar23 replied:
yes! is there a weekly/monthly call or is discord the best forum for discussions?

priscilawebdev replied:
yes you can always contact us on verdaccio discord chat 😃This week I'm a little absent, but surely the other members will be online to help you

agar23 replied:
sounds good 👍

priscilawebdev commented on src/components/PackageList/PackageList.tsx:
```diff
@@ -13,16 +13,16 @@ interface Props {
 
 export const PackageList: React.FC<Props> = props => {
-  const renderPackages: () => React.ReactElement<HTMLElement>[] = () => {
+  const renderPackages: () => ReactElement<HTMLElement>[] = () => {
```
Hm...I don't know if I agree with this type... maybe React.ReactNode?

agar23 replied:
done

priscilawebdev commented on src/components/RegistryInfoContent/RegistryInfoContent.test.tsx:
```diff
@@ -1,23 +1,24 @@
-    wrapper.setState({ tabPosition: 2 });
-    expect(wrapper.html()).toMatchSnapshot();
+  test('should load the appropiate tab content when the tab is clicked', () => {
+    const { container, getByTestId } = render(<RegistryInfoContent registryUrl={'http://localhost:4872'} scope={''} />);
```
I would keep the parameters in some object, because it's easy so do maintenance.

agar23 replied:
ohhh yes! done

priscilawebdev commented on src/components/RegistryInfoContent/RegistryInfoContent.test.tsx:
```diff
@@ -1,23 +1,24 @@
-    expect(wrapper.html()).toMatchSnapshot();
+  test('should load the appropiate tab content when the tab is clicked', () => {
+    const { container, getByTestId } = render(<RegistryInfoContent registryUrl={'http://localhost:4872'} scope={''} />);
+    expect(container.textContent).not.toContain('pnpm adduser --registry http://localhost:4872');
```
if you use an object as I mentioned above you could reuse the registryURL here :-)

agar23 replied:
done

priscilawebdev commented on src/components/RegistryInfoContent/RegistryInfoContent.tsx:
```diff
@@ -1,4 +1,4 @@
-import React, { Component } from 'react';
+import React, { useState } from 'react';
```
👏

agar23 replied:
😄

priscilawebdev commented on src/components/RegistryInfoContent/RegistryInfoContent.tsx:
```diff
@@ -11,86 +11,79 @@ import { default as Typography } from '../../muiComponents/Heading';
-
-  private handleChange = (event: React.ChangeEvent<{}>, tabPosition: number) => {
+export const RegistryInfoContent: React.FC<Props> = props => {
+  const [tabPosition, setTabPosition] = useState<State['tabPosition']>(0);
```
```suggestion
  const [tabPosition, setTabPosition] = useState(0);
```

Here you don't need to set a type, it will be automatically of type number, because of it's value, 0.

tmkn replied:
Nothing wrong with being explicit.
Setting a type is useful so that you can prevent a typo like this:
```typescript
const [tabPosition, setTabPosition] = useState('0'); //ok...
```
Without explicitly setting a type the compiler is fine with whatever you initialize the state with.
Now the type is `string` and any error because of type mismatch will show up where you consume the state, whilst if you would have been explicit:
```typescript
const [tabPosition, setTabPosition] = useState<number>('0'); //Argument of type '"0"' is not assignable to parameter of type 'number | (() => number)'
```
You would get an error right on this line, where the actual error happenend, since you **intended** the state to be of type `number` but (accidentally) passed in a `string` instead. :wink:

agar23 replied:
I agree although typescript is smart enough to infer the type here we should declare it. if not then we should remove our state interface.

priscilawebdev replied:
Hmmm I find it unnecessary, because we are not really setting this state so dynamically.... and even if we were passing a prop, it would be checked though of the interface or if we pass for example a string in the setTabPosition we will get a type error... so IMO is fine if you want to cast the state, but  it is very verbose and unnecessary.

agar23 replied:
understood

priscilawebdev commented on src/components/RegistryInfoContent/RegistryInfoContent.tsx:
```diff
@@ -11,86 +11,79 @@ import { default as Typography } from '../../muiComponents/Heading';
-  private renderNpmTab(scope: string, registryUrl: string): JSX.Element {
+  const renderNpmTab = (scope: string, registryUrl: string): JSX.Element => {
     return (
       <React.Fragment>
```
```suggestion
      <>
```

agar23 replied:
done

priscilawebdev commented on src/components/RegistryInfoContent/RegistryInfoContent.tsx:
```diff
@@ -11,86 +11,79 @@ import { default as Typography } from '../../muiComponents/Heading';
         <CopyToClipBoard text={getCLISetConfigRegistry(`${NODE_MANAGER.npm} set`, scope, registryUrl)} />
         <CopyToClipBoard text={getCLISetRegistry(`${NODE_MANAGER.npm} adduser`, registryUrl)} />
         <CopyToClipBoard text={getCLIChangePassword(NODE_MANAGER.npm, registryUrl)} />
       </React.Fragment>
```
```suggestion
      </>
```

agar23 replied:
done

priscilawebdev commented on src/components/RegistryInfoContent/RegistryInfoContent.tsx:
```diff
@@ -11,86 +11,79 @@ import { default as Typography } from '../../muiComponents/Heading';
-  private renderYarnTab(scope: string, registryUrl: string): JSX.Element {
+  const renderYarnTab = (scope: string, registryUrl: string): JSX.Element => {
     return (
       <React.Fragment>
```
```suggestion
      <>
```

agar23 replied:
done

priscilawebdev commented on src/components/RegistryInfoContent/RegistryInfoContent.tsx:
```diff
@@ -11,86 +11,79 @@ import { default as Typography } from '../../muiComponents/Heading';
     return (
       <React.Fragment>
         <CopyToClipBoard text={getCLISetConfigRegistry(`${NODE_MANAGER.yarn} config set`, scope, registryUrl)} />
       </React.Fragment>
```
```suggestion
      </>
```

agar23 replied:
done

priscilawebdev commented on src/components/RegistryInfoContent/RegistryInfoContent.tsx:
```diff
@@ -11,86 +11,79 @@ import { default as Typography } from '../../muiComponents/Heading';
-  private renderPNpmTab(scope: string, registryUrl: string): JSX.Element {
+  const renderPNpmTab = (scope: string, registryUrl: string): JSX.Element => {
     return (
       <React.Fragment>
```
```suggestion
      <>
```

agar23 replied:
done

priscilawebdev commented on src/components/RegistryInfoContent/RegistryInfoContent.tsx:
```diff
@@ -11,86 +11,79 @@ import { default as Typography } from '../../muiComponents/Heading';
+  };
 
-  private renderPNpmTab(scope: string, registryUrl: string): JSX.Element {
+  const renderPNpmTab = (scope: string, registryUrl: string): JSX.Element => {
```
```suggestion
  const renderPnpmTab = (scope: string, registryUrl: string): JSX.Element => {
```

agar23 replied:
done

priscilawebdev commented on src/components/RegistryInfoContent/RegistryInfoContent.tsx:
```diff
@@ -11,86 +11,79 @@ import { default as Typography } from '../../muiComponents/Heading';
         <CopyToClipBoard text={getCLISetConfigRegistry(`${NODE_MANAGER.pnpm} set`, scope, registryUrl)} />
         <CopyToClipBoard text={getCLISetRegistry(`${NODE_MANAGER.pnpm} adduser`, registryUrl)} />
         <CopyToClipBoard text={getCLIChangePassword(NODE_MANAGER.pnpm, registryUrl)} />
       </React.Fragment>
```
```suggestion
      </>
```

agar23 replied:
done

priscilawebdev commented on src/components/RegistryInfoContent/index.ts:
```diff
@@ -1 +1 @@
-export { default } from './RegistryInfoContent';
```
please try to keep the standard 😉

agar23 replied:
same as https://github.com/verdaccio/ui/pull/229#discussion_r339846597 wondering why we want to use default exports?

priscilawebdev commented on src/components/RegistryInfoContent/RegistryInfoContent.tsx:
```diff
@@ -11,86 +11,79 @@ import { default as Typography } from '../../muiComponents/Heading';
+    const { scope, registryUrl } = props;
+
+    return (
+      <React.Fragment>
```
```suggestion
      <>
```

agar23 replied:
done

priscilawebdev commented on src/components/RegistryInfoContent/RegistryInfoContent.tsx:
```diff
@@ -11,86 +11,79 @@ import { default as Typography } from '../../muiComponents/Heading';
+        {tabPosition === 0 && <TabContainer>{renderNpmTab(scope, registryUrl)}</TabContainer>}
+        {tabPosition === 1 && <TabContainer>{renderPNpmTab(scope, registryUrl)}</TabContainer>}
+        {tabPosition === 2 && <TabContainer>{renderYarnTab(scope, registryUrl)}</TabContainer>}
+      </React.Fragment>
```
```suggestion
      </>
```

agar23 replied:
done

agar23 replied:
done

priscilawebdev commented on src/components/PackageList/PackageList.tsx:
```diff
@@ -13,16 +13,16 @@ interface Props {
 }
 
 export const PackageList: React.FC<Props> = props => {
```
Please destruct the props here, then you don't need to to like this: 
![image](https://user-images.githubusercontent.com/29228205/67668964-3d811080-f971-11e9-8598-90029878c917.png)

agar23 replied:
we are not destructuring the props for this component.

priscilawebdev replied:
Thank you!! :-)

priscilawebdev replied:
I would still destruct the props like this: `const PackageList: React.FC<Props> =({ packages }) => `

agar23 replied:
ahhh I see what you mean! done.

priscilawebdev commented:
@agar23 thank you very much for your contributing 🙂👏I would apply just a few minor changes ... please check my comments :-)

agar23 commented:
> @agar23 thank you very much for your contribution 🙂👏I would apply just a few minor changes ... please check my comments :-)

np, I addressed most of the comments. but had a couple of questions/comments I left.

priscilawebdev commented on src/components/PackageList/PackageList.tsx:
```diff
@@ -13,21 +13,22 @@ interface Props {
-    return props.packages.map((pkg, i) => {
+  const { packages } = props;
+  const renderPackages: () => ReactNode[] = () => {
+    return packages.map((pkg, i) => {
```
This is only a suggestion, but I would avoid creating another constant by doing this:
```suggestion
    return packages.map(({ name, version, description, time, keywords, dist, homepage, bugs, author }, i) => {
```

agar23 replied:
ohh yes! done

priscilawebdev commented on src/components/PackageList/PackageList.tsx:
```diff
@@ -12,22 +12,23 @@ interface Props {
+    return packages.map(({ name, version, description, time, keywords, dist, homepage, bugs, author, license }, i) => {
       // TODO: move format license to API side.
-      const license = formatLicense(pkg.license);
+      const _license = formatLicense(license);
```
Please why did you add `_` ?

agar23 replied:
we cant name this variable `license` the props already have a prop named `license` I presume this why the component was not destructuring the props before. this way we could call `pkg.license`

priscilawebdev commented on src/components/RegistryInfoContent/RegistryInfoContent.tsx:
```diff
@@ -11,86 +11,77 @@ import { default as Typography } from '../../muiComponents/Heading';
-  );
-}
+const RegistryInfoContent: React.FC<Props> = props => {
+  const [tabPosition, setTabPosition] = useState<State['tabPosition']>(0);
```
```suggestion
  const [tabPosition, setTabPosition] = useState(0);
```

agar23 replied:
as discussed I think it should be ok to be explicit here

priscilawebdev commented:
@agar23 almost there :-)

juanpicado commented:
@agar23 great job, there are one conflict and seems linting is failing :) looking forward to merge it soon :)
